### PR TITLE
Rough first version

### PR DESCRIPTION
**Pull request by jovankomatovic**
This is just the rough first version.
I have used some functions to signal that some event has happened, i.e., function "OnEvent" is invoked once an event "Event" occurs.

### CODE REVIEW

File: spec/valset-update-protocol.md [hunk 1]
```diff
@@ -18,6 +26,16 @@ struct ValSetUpdateAcknowledgement {
 }
 ```
 
+`EvidencePacket` is sent by the daughter chain once an evidence of misbehaviour is collected or an unbonding period is over at the daughter chain.
+
+```golang
+struct EvidencePacket {
+  validator  Validator
+  // is one evidence enough?
```
Comment (josef-widder) [inline spec/valset-update-protocol.md:34]: We have to discuss what this evidence will look like. Perhaps we don't even have to submit precise evidence, given that the daughter chain might not be able to validate it (completely) anyhow. Perhaps we trust the daughter chain to only send valid accusations? Or just a hash (proof) of the evidence.

File: spec/valset-update-protocol.md [hunk 2]
```diff
@@ -139,24 +157,66 @@ func sendValSetUpdatePacket(
...
-  newValSet = data.valSet // how we pass this information to Tendermint?
-  return ack
+  Type type = packet.type()
+
+  switch (type) {
+    // executed on the daughter chain
+    // validator set change demanded by the mother chain
+    case ValSetUpdatePacket:
+      ValSetUpdatePacket data = packet.data
+      // perform some checks?
```
Comment (josef-widder) [inline spec/valset-update-protocol.md:181]: We have to understand what are the guarantees regarding a packet that reaches this point.

File: spec/valset-update-protocol.md [hunk 3]
```diff
@@ -139,24 +157,66 @@ func sendValSetUpdatePacket(
...
+
+      // construct default acknowledgement of success
+      ValSetUpdateAcknowledgement = ValSetUpdateAcknowledgement{true, null}
+
+      // who manages validator sets in the SDK app and how do we pass this info to Tendermint?
+      newValSet = data.valSet // how do we pass this information to Tendermint?
```
Comment (josef-widder) [inline spec/valset-update-protocol.md:187]: I guess we have to set up and define a future cross chain validation module or plan how to extend the existing staking module

**Review by josef-widder — APPROVED** [review]
For the functions it would be good to write preconditions and postconditions like in [verification](https://github.com/tendermint/spec/blob/master/rust-spec/lightclient/verification/verification_001_published.md#lcv-func-valid1).  
One question I have is that ICS 20 (which was the role model) is purely message-driven. It is reactive. Why for cross-chain, we might need to be active (I the timeout for the unbonding period is reached, we need to send a packet). I am not sure how to write this reactive logic.

File: spec/valset-update-protocol.md (hunk 1 above)
Comment (jovankomatovic) [reply spec/valset-update-protocol.md:34]: If the mother chain does not check the evidence of a misbehavior, could we have a situation where faulty validators collude and accuse a correct process?

File: spec/valset-update-protocol.md (hunk 2 above)
Comment (jovankomatovic) [reply spec/valset-update-protocol.md:181]: I agree.

File: spec/valset-update-protocol.md (hunk 3 above)
Comment (jovankomatovic) [reply spec/valset-update-protocol.md:187]: Yes, I agree. Maybe we could discuss this, so that we know how to proceed.

**Comment by jovankomatovic** [comment]
I have added preconditions and postconditions, as @josef-widder suggested. Also, I have introduced some minor improvements.

**Review by cwgoes — COMMENTED** [review]
Thanks for the draft. I'm actually not sure if the ICS spec format is the best way to start specifying this protocol - it's fine if you want to, but if something more formal is better/easier, that's fine too, we can even change the ICS format potentially, it's just a draft (also see https://github.com/cosmos/ics/issues/468, sadly not done yet)

File: spec/valset-update-protocol.md (hunk 1 above)
Comment (josef-widder) [reply spec/valset-update-protocol.md:34]: I guess so. We should discuss possible scenarios, and what we can do in each of them.

**Comment by josef-widder** [comment]
I wasn't aware of the discussion in https://github.com/cosmos/ics/issues/468. Great ideas! Perhaps we can write this crosschain validation specification in a form that could serve as a template for the rewrite of the ICSs?

Adding the preconditions and postconditions already makes it clearer (and makes a big step towards modeling in TLA+).

What is missing in the ICSs is to make explicit who calls what function and when and under what conditions. Also, what would a perfect relayer do: what to read from DC and what to submit to MC etc. I am not sure about the format. I will think about this a bit until tomorrow ;-)

File: spec/valset-update-protocol.md (hunk 1 above)
Comment (milosevic) [reply spec/valset-update-protocol.md:34]: The philosophy of IBC is not to trust counterparty chain. So we probably can't rely blindly on daughter chain to execute evidence protocol honestly. I guess this is probably one of the reasons Chris calls for direct evidence submission to a mother chain. Having overlapping validator sets might potentially help, so we might need to have different protocol versions for different scenarios.

File: spec/valset-update-protocol.md [hunk 4]
```diff
@@ -18,6 +41,16 @@ struct ValSetUpdateAcknowledgement {
 }
 ```
 
+`UnbondingTimeoutExpiredPacket` is sent by the daughter chain to the mother chain to inform the mother chain
+that the unbonding period for a specific validator has expired.
+
+```golang
+struct UnbondingTimeoutExpiredPacket {
+  validator Validator
```
[the remark this answers is not in the document]
Comment (jovankomatovic) [reply spec/valset-update-protocol.md:49]: Yes, I agree. Will take care of this.

File: spec/valset-update-protocol.md [hunk 5]
```diff
@@ -139,24 +172,60 @@ func sendValSetUpdatePacket(
 }
 ```
 
+```golang
+func sendUnbondingTimeoutExpiredPacket(
```
[the remark this answers is not in the document]
Comment (jovankomatovic) [reply spec/valset-update-protocol.md:176]: I agree. Will do that separation.

File: spec/valset-update-protocol.md [hunk 6]
```diff
@@ -139,24 +172,60 @@ func sendValSetUpdatePacket(
...
+  validator: Validator) {
```
[the remark this answers is not in the document]
Comment (jovankomatovic) [reply spec/valset-update-protocol.md:177]: I specified it below. This should be invoked once there is a block committed on the daughter chain with some time T. Moreover, suppose that unbonding is started at some time U. Then, the function is triggered if T - U > unbonding periods (3 weeks). Does this make sense?

**Comment by josef-widder** [comment]
should we merge and start a fresh PR from here?

**Comment by jovankomatovic** [comment]
Could you give me one more day to separate the functions into mother and daughter chain specific?
And to clean the write-up a bit?

**Comment by josef-widder** [comment]
sure. it was just a suggestion ;-)

**Comment by jovankomatovic** [comment]
I have separated functions into two sections. I think we can merge it and then I will add data flow and solve potential problems. Thanks!

**Pull request merged**